doc/spec/debug-ros-nodes.md

```diff
@@ -0,0 +1,269 @@
+Our extension automatically chooses `cppvsdbg` for MSVC debugger when it's running on Windows, and `cppdbg` for gdb-based debugger otherwise.
+
+### Attach into a Python process
+
+Our extension uses [`ptvsd`][ptvsd] to enable debugging Python scripts.
+Generally, a Python script needs to be launched through the `ptvsd` module to be able to be debugged:
+
+```bash
+python -m ptvsd –host <host> --port <port> <script>
+```
+
+Once a script is launched like this, a Python attach session could be launched inside VS Code to attach to the `ptvsd` server running inside the process.
+
+However, Python processes launched by ROS tools (`rosrun` and `roslaunch`) are usually launched natively.
+To debug a Python process without a running `ptvsd` instance, a `ptvsd` instance needs to be injected into the process first with the `--pid <pid>` flag.
+Therefore, for normal Python processes in ROS, the workflow would be:
+
+1. get the PID of the Python process (we reuse the `processPicker` module from `vscode-cpptools`)
+2. inject `ptvsd` using API exposed by `vscode-python` (this avoids potential `ptvsd` version mismatch)
+3. start a Python-attach debug session with `vscode-python`
+
+### Example for `attach`
+
+```json
+{
+    "configurations": [
+        {
+            "name": "ROS: Attach",
+            "type": "ros",
+            "request": "attach"
+        },
+        {
+            "name": "ROS: Attach to Python",
+            "type": "ros",
+            "request": "attach",
+            "runtime": "Python"
+        }
+    ]
+}
+```
+
+### Note on attach-debug
+
+Microsoft's VS Code extensions for [C++][ms-vscode.cpptools] and [Python][ms-python.python] provide flexible mechanisms of attaching into a [C++][vscode_cpp_debug] or [Python][vscode_python_debug] process.
+This extension only aims to enable a basic and generic attach-debug flow.
+Please use language-specific extensions for any specific debugging requirements including:
+
+* specifying `symbolSearchPath` or `miDebuggerPath` for C++ debugging
+* attaching to and debugging a remote process
+
+## Launch
+
+Our goal for launch-debug is to mimic the experience of debugging a single program in VS Code.
+When a `.launch` file is launched in debug mode, we want to launch all the ROS nodes defined in it in debug mode at the same time.
+Similar to attach-debug, we want to utilize all the language-specific debugging functionalities provided by `vscode-python` and `vscode-cpptools`.
+
+![launch_debug][launch_debug]
+
+From another perspective, we aim to translate launch configurations in a `.launch` file into debug configurations for VS Code that will be executed all at the same time.
+To do this, we use [the command line functionality of `roslaunch`][roslaunch_commandline] to extract launch requests from a `.launch` file:
+
+1. get parameters defined in the `.launch` file with `roslaunch --dump-params <launch-file>`
```

Comment:
It would be great to see what's the gap between ROS2 and ROS1 when we try to replicate the same model.

Reply:
planning to add section about ROS2 when we start that work.